### PR TITLE
Rescan and fetch blocks

### PR DESCRIPTION
**Pull request by rorp**

### CODE REVIEW

File: chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala [hunk 1]
```diff
@@ -429,12 +431,16 @@ case class ChainHandler(
...
           val startHeight = end - (batchSize - 1)
           val endHeight = end
-          val newAcc: Future[Vector[DoubleSha256DigestBE]] = for {
+          val newAcc = for {
             compactFilterDbs <- filterDAO.getBetweenHeights(startHeight,
                                                             endHeight)
-            groupSize = calcGroupSize(compactFilterDbs.size)
-            grouped = compactFilterDbs.grouped(groupSize)
-            filtered <- findMatches(grouped)
+            filtered <- if (compactFilterDbs.isEmpty)
```
Comment (Christewart) [inline chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala]: Do we really need this check here? Can't we just delegate this to the else clause and have `findMatches()` not work on a `Vector.empty`?

File: node-test/src/test/scala/org/bitcoins/node/NeutrinoNodeTest.scala [hunk 2]
```diff
@@ -90,33 +121,68 @@ class NeutrinoNodeTest extends NodeUnitTest {
       startGenF.flatMap { _ =>
         //we should expect 5 headers have been announced to us via
         //the send headers message.
-        def has6BlocksF =
+        def hasBlocksF =
           RpcUtil.retryUntilSatisfiedF(conditionF = () => {
             node
               .chainApiFromDb()
-              .flatMap(_.getBlockCount.map(_ == 6))
+              .flatMap(_.getBlockCount.map(_ == 113))
```
Comment (Christewart) [inline node-test/src/test/scala/org/bitcoins/node/NeutrinoNodeTest.scala]: Can we mave `113` to something like 

```scala
val EXPECTED_BLOCK_COUNT = 113
```

File: node/src/main/scala/org/bitcoins/node/Node.scala [hunk 3]
```diff
@@ -160,14 +162,16 @@ trait Node extends P2PLogger {
     }
   }
 
+  def onStop(): Future[Unit]
```
Comment (Christewart) [inline node/src/main/scala/org/bitcoins/node/Node.scala]: What are these? Scaladoc please

File: node/src/main/scala/org/bitcoins/node/Node.scala [hunk 4]
```diff
@@ -176,9 +180,11 @@ trait Node extends P2PLogger {
     } yield {
       peerMsgSenderF.map(_.sendGetHeadersMessage(hash.flip))
       logger.info(s"Starting sync node, height=${header.height} hash=$hash")
-    }
+    }).flatMap(_ => onSync())
   }
 
+  def onSync(): Future[Unit]
```
Comment (Christewart) [inline node/src/main/scala/org/bitcoins/node/Node.scala]: What is this? Scaladoc please.

File: node/src/main/scala/org/bitcoins/node/Node.scala [hunk 5]
```diff
@@ -176,9 +180,11 @@ trait Node extends P2PLogger {
     } yield {
       peerMsgSenderF.map(_.sendGetHeadersMessage(hash.flip))
       logger.info(s"Starting sync node, height=${header.height} hash=$hash")
-    }
+    }).flatMap(_ => onSync())
```
Comment (Christewart) [inline node/src/main/scala/org/bitcoins/node/Node.scala]: I don't think this actually guarantees this callback is called when we are synced. So this is a little misleading IMO. Basically this should be called `onStartSync()`.

File: node/src/main/scala/org/bitcoins/node/SpvNode.scala [hunk 6]
```diff
@@ -76,4 +76,7 @@ case class SpvNode(
     }
   }
 
+  override def onStop(): Future[Unit] = Future.successful(())
```
Comment (Christewart) [inline node/src/main/scala/org/bitcoins/node/SpvNode.scala]: nit: `FutureUtil.unit`

File: node-test/src/test/scala/org/bitcoins/node/NeutrinoNodeTest.scala [hunk 7]
```diff
@@ -90,33 +121,68 @@ class NeutrinoNodeTest extends NodeUnitTest {
...
-          _ <- has6FiltersF
+          _ <- hasBlocksF
+          _ <- hasFilterHeadersF
+          _ <- hasFiltersF
         } yield succeed
       }
   }
 
+  it must "download a block that matches a compact block filter" taggedAs (UsesExperimentalBitcoind) in {
+    nodeConnectedWithBitcoind: NeutrinoNodeFundedWalletBitcoind =>
```
Comment (Christewart) [inline node-test/src/test/scala/org/bitcoins/node/NeutrinoNodeTest.scala:152]: Let's start using the bitcoin-s wallet when we can!

```scala
     val node = nodeConnectedWithBitcoind.node
      val bitcoind = nodeConnectedWithBitcoind.bitcoindRpc
      val bitcoinsWallet = nodeConnectedWithBitcoind.wallet
      var cancelable: Option[Cancellable] = None

      for {
        addressFromBitcoins <- bitcoinsWallet.getNewAddress()
        _ = addressFromBitcoindP.success(addressFromBitcoins)
        ...
```

**Review by Christewart — CHANGES_REQUESTED** [review]

File: chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala (hunk 1 above)
Comment (rorp) [reply chain/src/main/scala/org/bitcoins/chain/blockchain/ChainHandler.scala]: We do. `Iterable.grouped()` throws if group size is less than 1. Also there's no reason to allocate so many entities for an empty DB response.

**Pull request merged**